src/transfermatrix/transfermatrix.jl

```diff
@@ -0,0 +1,69 @@
+abstract type AbstractTransferMatrix end;
+
+# single site transfer
+struct SingleTransferMatrix{A<:AbstractTensorMap,B,C<:AbstractTensorMap} <: AbstractTransferMatrix
+    above::A
+    middle::B
+    below::C
+    isflipped::Bool
+end
+
+#the product of transfer matrices is its own type
+struct ProductTransferMatrix{T<:AbstractTransferMatrix} <: AbstractTransferMatrix
+    tms :: Vector{T} # I don't want to use tuples, as an infinite mps transfer matrix will then be non-inferable
+end
+
+# a subset of possible operations, but certainly not all of them
+Base.:*(prod::ProductTransferMatrix{T},tm::T) where T <:AbstractTransferMatrix = ProductTransferMatrix(vcat(prod.tms,tm));
+Base.:*(tm::T,prod::ProductTransferMatrix{T}) where T <:AbstractTransferMatrix = ProductTransferMatrix(vcat(prod.tms,tm));
+Base.:*(tm1::T,tm2::T) where T <: SingleTransferMatrix = ProductTransferMatrix([tm1,tm2]);
+
+# regularized transfer matrices; where we project out after every full application
+struct RegTransferMatrix{T<:AbstractTransferMatrix,L,R} <: AbstractTransferMatrix
+    tm::T
+    lvec::L
+    rvec::R
+end
+
+#flip em
+TensorKit.flip(tm::SingleTransferMatrix) = SingleTransferMatrix(tm.above,tm.middle,tm.below,tm.isflipped ⊻ true);
+TensorKit.flip(tm::ProductTransferMatrix) = ProductTransferMatrix(flip.(reverse(tm.tms)));
+TensorKit.flip(tm::RegTransferMatrix) = RegTransferMatrix(flip(tm.tm),tm.rvec,tm.lvec);
+
+# TransferMatrix acting on a vector using *
+Base.:*(tm::AbstractTransferMatrix,vec) = tm(vec);
+Base.:*(vec,tm::AbstractTransferMatrix) = flip(tm)(vec);
+
+# TransferMatrix acting as a function
+(d::ProductTransferMatrix)(vec) = foldr((a,b)->a(b),d.tms,init=vec);
+(d::SingleTransferMatrix)(vec) = d.isflipped  ? transfer_left(vec,d.middle,d.above,d.below) : transfer_right(vec,d.middle,d.above,d.below);
+(d::RegTransferMatrix)(vec) = regularize!(d.tm*vec,d.lvec,d.rvec);
+
+
+# constructors
+TransferMatrix(a) = TransferMatrix(a,nothing,a);
+TransferMatrix(a,b) = TransferMatrix(a,nothing,b);
+TransferMatrix(a,b,c) = TransferMatrix(a,b,c,false);
+TransferMatrix(a::AbstractTensorMap,b,c::AbstractTensorMap,isflipped) = SingleTransferMatrix(a,b,c,isflipped);
+function TransferMatrix(a::AbstractVector,b,c::AbstractVector,isflipped)
```

Comment:
`ProductTransferMatrix` only accepts `Vector`; I assume the conversion will be implicit or a consequence of broadcasting. I keep wondering whether something like `StaticVector` should be allowed, which does encode length information (and probably preserves it in broadcasting). But for now the choice seems to be only `Vector`, so I would also enforce this here, unless there was some other reason.

Reply:
this would be tricky, as mpskit often deals with periodicarrays, for which transfermatrix would then fail. It currently works indeed because of the broadcasting, which turns it into a vector.

I can also make producttransfermatrix accept any AbstractVector subtype?